Convert single quotes when escaping HTML entities

hemberger opened the pull request:
**NOTICE**: This PR introduces a change in behavior. We may want to carefully consider our options before merging, as there are compelling reasons for each possible solution to this issue. Thank you!

In PHP 8.1, the default flags for `htmlspecialchars` have changed so
that single quotes are now escaped (for security reasons). As a result,
we must explicitly pass flags to `htmlspecialchars` in order for NBBC
to have consistent behavior between pre-8.1 and post-8.1 PHP versions.

If we wanted to retain the pre-8.1 behavior, then we would use the
ENT_COMPAT flag. However, since it seems that the PHP devs made this
change for a good reason, it's likely worth changing the behavior of
NBBC as well. Therefore, this PR chooses to pass ENT_QUOTES instead
(the same as ENT_COMPAT, except also escapes single quotes).

Note that ENT_QUOTES is not quite the 8.1 default for htmlspecialchars,
which also includes ENT_SUBSTITUTE (for security reasons again). This
begins to be a bit cumbersome, but could be included for full parity
with PHP 8.1, if desired.

Added CircleCI tests for PHP 8.1, which now pass. Tests were updated to
reflect single quotes being escaped as `&#039;`, and also added a test
case for `setEscapeContent(false)`, which was previously untested.

This site has a thorough summary of the change to `htmlspecialchars`:
https://php.watch/versions/8.1/html-entity-default-value-changes

hemberger committed:
Convert single quotes when escaping HTML entities

In PHP 8.1, the default flags for `htmlspecialchars` have changed so
that single quotes are now escaped (for security reasons). As a result,
we must explicitly pass flags to `htmlspecialchars` in order for NBBC
to have consistent behavior between pre-8.1 and post-8.1 PHP versions.

If we wanted to retain the pre-8.1 behavior, then we would use the
ENT_COMPAT flag. However, since it seems that the PHP devs made this
change for a good reason, it's likely worth changing the behavior of
NBBC as well. Therefore, this PR chooses to pass ENT_QUOTES instead
(the same as ENT_COMPAT, except also escapes single quotes).

Note that ENT_QUOTES is not quite the 8.1 default for htmlspecialchars,
which also includes ENT_SUBSTITUTE (for security reasons again). This
begins to be a bit cumbersome, but could be included for full parity
with PHP 8.1, if desired.

Added CircleCI tests for PHP 8.1, which now pass. Tests were updated to
reflect single quotes being escaped as `&#039;`, and also added a test
case for `setEscapeContent(false)`, which was previously untested.

This site has a thorough summary of the change to `htmlspecialchars`:
https://php.watch/versions/8.1/html-entity-default-value-changes

tburry commented:
Does this just affect the output or also the input. Like can I still use `[tag attr='something']` or is that gone?

hemberger committed:
Add test cases for single quoted tag values

Neither double quotes nor single quotes should not be escaped when
they are used inside of a tag. That is, the following should all
result in the same HTML:

* [tag=value param=value]
* [tag='value' param='value']
* [tag="value" param="value"]

hemberger commented:
> Does this just affect the output or also the input. Like can I still use `[tag attr='something']` or is that gone?

@tburry An excellent question! I've added explicit test cases to verify that the following all result in the same HTML:

* `[tag=value param=value]`
* `[tag='value' param='value']`
* `[tag="value" param="value"]`

That is, you can still quote tag and parameter values with either single or double quotes. (We do not exhaustively test every tag that allows values or parameter values, but I assume it is sufficient to spot check this.)

tburry reviewed (approved):
Great. Thanks so much for such a thorough PR.
